File: core/internal/activity/kfa.go
// Package activity вычисляет группу коэффициента физической активности
// (КФА) по ответам мини-анкеты.
//
// Источник правил: docs/АНКЕТА_активности.md, дерево решений без Q2/Q5
// (зафиксированное решение по открытому вопросу №1).
//
// Анкета содержит три вопроса:
//   Q1 — характер обычной дневной активности;
//   Q3 — частота отдельной физической нагрузки длительностью не менее 30 мин;
//   Q4 — интенсивность этой нагрузки (nullable, не задаётся при Q3=none).
package activity

import (
	"fmt"

	"nutrition-core/internal/domain"
)

type Q1 string

const (
	Q1Sedentary         Q1 = "sedentary"
	Q1StandingLow       Q1 = "standing_low"
	Q1FrequentMovement  Q1 = "frequent_movement"
	Q1HeavyPhysical     Q1 = "heavy_physical"
)

type Q3 string

const (
	Q3None    Q3 = "none"
	Q3OneTwo  Q3 = "1_to_2"
	Q3ThreeFive Q3 = "3_to_5"
	Q3SixPlus Q3 = "6_plus"
)

type Q4 string

const (
	Q4None     Q4 = ""
	Q4Light    Q4 = "light"
	Q4Moderate Q4 = "moderate"
	Q4Intense  Q4 = "intense"
)

// Survey — нормализованные ответы пользователя.
type Survey struct {
	Q1 Q1
	Q3 Q3
	Q4 Q4 // пустая, если Q3 == none
}

// DeriveKFA реализует дерево решений по docs/АНКЕТА_активности.md.
// Решение по открытому вопросу №1 — вариант B (без Q2/Q5):
//
//   IV  = Q1=heavy_physical OR (Q3=6_plus AND Q4=intense)
//   III = Q1=frequent_movement OR (Q3=3_to_5 AND Q4 in {moderate, intense})
//   II  = Q1=standing_low
//          OR Q3=1_to_2
//          OR (Q3 in {3_to_5, 6_plus} AND Q4=light)
//   I   = иначе
//
// Порядок проверки сверху вниз — попадание в более высокую группу
// фиксируется первым.
func DeriveKFA(s Survey) (domain.KfaGroup, error) {
	if err := s.validate(); err != nil {
		return "", err
	}

	// IV
	if s.Q1 == Q1HeavyPhysical {
		return domain.KfaIV, nil
	}
	if s.Q3 == Q3SixPlus && s.Q4 == Q4Intense {
		return domain.KfaIV, nil
	}

	// III
	if s.Q1 == Q1FrequentMovement {
		return domain.KfaIII, nil
	}
	if s.Q3 == Q3ThreeFive && (s.Q4 == Q4Moderate || s.Q4 == Q4Intense) {
		return domain.KfaIII, nil
	}

	// II
	if s.Q1 == Q1StandingLow {
		return domain.KfaII, nil
	}
	if s.Q3 == Q3OneTwo {
		return domain.KfaII, nil
	}
	if (s.Q3 == Q3ThreeFive || s.Q3 == Q3SixPlus) && s.Q4 == Q4Light {
		return domain.KfaII, nil
	}

	return domain.KfaI, nil
}

func (s Survey) validate() error {
	switch s.Q1 {
	case Q1Sedentary, Q1StandingLow, Q1FrequentMovement, Q1HeavyPhysical:
	default:
		return fmt.Errorf("activity: invalid Q1 %q", s.Q1)
	}
	switch s.Q3 {
	case Q3None, Q3OneTwo, Q3ThreeFive, Q3SixPlus:
	default:
		return fmt.Errorf("activity: invalid Q3 %q", s.Q3)
	}
	if s.Q3 == Q3None {
		if s.Q4 != Q4None {
			return fmt.Errorf("activity: Q4 must be empty when Q3=none, got %q", s.Q4)
		}
		return nil
	}
	switch s.Q4 {
	case Q4Light, Q4Moderate, Q4Intense:
	default:
		return fmt.Errorf("activity: Q4 required when Q3!=none, got %q", s.Q4)
	}
	return nil
}

File: web/src/lib/kfa.ts
// Дерево решений «ответы анкеты → группа КФА» по docs/АНКЕТА_активности.md
// (вариант B — без Q2/Q5; в источниках обосновано через WHO 2020 + EFSA 2013).
// Параллельная копия core/internal/activity/kfa.go — фронт должен уметь
// считать КФА офлайн (PWA), поэтому дублирование оправдано.

import type { Kfa } from "../api/types";

export type Q1 = "sedentary" | "standing_low" | "frequent_movement" | "heavy_physical";
export type Q3 = "none" | "1_to_2" | "3_to_5" | "6_plus";
export type Q4 = "" | "light" | "moderate" | "intense";

export interface Survey {
  q1: Q1;
  q3: Q3;
  q4: Q4; // пусто, если q3 == "none"
}

export function deriveKfa(s: Survey): Kfa {
  // IV
  if (s.q1 === "heavy_physical") return "IV";
  if (s.q3 === "6_plus" && s.q4 === "intense") return "IV";

  // III
  if (s.q1 === "frequent_movement") return "III";
  if (s.q3 === "3_to_5" && (s.q4 === "moderate" || s.q4 === "intense")) return "III";

  // II
  if (s.q1 === "standing_low") return "II";
  if (s.q3 === "1_to_2") return "II";
  if ((s.q3 === "3_to_5" || s.q3 === "6_plus") && s.q4 === "light") return "II";

  // I по умолчанию
  return "I";
}

// Тексты вариантов ответов из АНКЕТА_активности.md.
export const Q1_OPTIONS: { value: Q1; label: string; hint?: string }[] = [
  { value: "sedentary",         label: "В основном сижу",
    hint: "работа за компьютером, учёба, длительное пребывание дома без активного движения" },
  { value: "standing_low",      label: "В основном стою, но мало перемещаюсь",
    hint: "парикмахер, продавец за стойкой, преподаватель у доски" },
  { value: "frequent_movement", label: "Часто хожу или перемещаюсь",
    hint: "официант, продавец в торговом зале, медработник, курьер" },
  { value: "heavy_physical",    label: "Выполняю физически тяжёлую работу",
    hint: "склад, стройка, производство, грузовые работы" },
];

export const Q3_OPTIONS: { value: Q3; label: string }[] = [
  { value: "none",     label: "Нет таких нагрузок" },
  { value: "1_to_2",   label: "1–2 раза в неделю" },
  { value: "3_to_5",   label: "3–5 раз в неделю" },
  { value: "6_plus",   label: "6 и более раз в неделю" },
];

export const Q4_OPTIONS: { value: Exclude<Q4, "">; label: string; hint?: string }[] = [
  { value: "light",    label: "Лёгкая",
    hint: "спокойная зарядка, растяжка, неспешная йога, прогулка в умеренном темпе" },
  { value: "moderate", label: "Умеренная",
    hint: "фитнес, плавание, велосипед, быстрая ходьба" },
  { value: "intense",  label: "Интенсивная",
    hint: "силовая, бег, интервальные тренировки, единоборства" },
];
